drivers: serial: stm32: Refactor for PM handling

knthm opened the pull request:
### Summary

This PR refactors the initialization/configuration logic of STM32 UART driver. It is intended to complement PR #59200 by making reinitialization of the peripheral clocks and registers more straightforward, for when the SoC loses their context and contents in certain low-power modes.

In refactoring I've also stumbled on a few bugs that I've fixed.

- Added static `uart_config` struct to `STM32_UART_INIT` macro
- Added `uart_config` struct pointer as `uart_stm32_config` member
- (Re)moved now superfluous `baud_rate`, `parity`, and `hw_flow_control` struct members
- Moved overlapping baudrate/parity/data-/stop-bits/flow control configuration to common function, with persistence in static instance `uart_config` struct
- Removed incorrect runtime parity/data bits configuration, and replaced with thoroughly tested compile-time `BUILD_ASSERT` statements
  - Original runtime logic in `init` forced a very small subset of UART configurations, and straight-up ignored many valid DT property combinations (e.g 7E1)
  - This now also allows invalid configurations to get flagged compile-time
  - The previous logic partially conflicted with the logic in `uart_stm32_configure`, especially that relating to parity/data-bits combinations, where either enforced opposite configurations
- The `#if defined(LL_USART_STOPBITS_1_5) && HAS_LPUART_1` conditional broke "half-bit" parity application configurations of USARTs, this is now fixed


### Change Description

Add static uart_config struct to init macro with corresponding pointer as uart_stm32_config struct member. This struct will store boot and runtime UART configuration for reinitialization upon exiting low-power modes, where register contents might be lost.

Remove hw_flow_control, parity, and baud_rate from uart_stm32_config and uart_stm32_data structs, respectively, as the need for these is now obviated by the presence of the uart_config struct, which contains equivalent members.

Add default baudrate, parity, stop, and data bits constants, used to initialize the uart_config struct, in case the corresponding UART DT properties are not set.

Move overlapping UART parameter configuration to new uart_stm32_parameters_set function. This function is called by uart_stm32_configure upon application/subsystem configuration, and uart_stm32_registers_configure upon (re-)initialization of driver instances.

Fix flawed stop bit/LPUART conditionals, which didn't allow for 0.5/1.5 stop bit USART configurations via application uart_configure syscalls.

Replace incorrect and limited runtime parity/data bits conditional configuration with build-time logic using BUILD_ASSERT. This allows for more flexible UART configurations, while preventing invalid DTS configurations at build-time.

### Issues Closed by this PR

- #56003

### Related Issues or Pull Requests

- #59200

knthm committed:
drivers: serial: stm32: Add persistent uart_config struct

Add static uart_config struct to init macro with corresponding pointer
as uart_stm32_config struct member. This struct will store boot and
runtime UART configuration for reinitialization upon exiting low-power
modes, where register contents might be lost.

Remove hw_flow_control, parity, and baud_rate from uart_stm32_config and
uart_stm32_data structs, respectively, as the need for these is now
obviated by the presence of the uart_config struct, which contains
equivalent members.

Add default baudrate, parity, stop, and data bits constants, used to
initialize the uart_config struct, in case the corresponding UART DT
properties are not set.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: serial: stm32: Refactor for PM handling

Move clock enable and register configuration to their own functions in
preparation for later uart_stm32_reinit function that will use these in
addition to uart_stm32_init.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: serial: stm32: Consolidate UART config

Move overlapping UART parameter configuration to new
uart_stm32_parameters_set function. This function is called by
uart_stm32_configure upon application/subsystem configuration, and
uart_stm32_registers_configure upon (re-)initialization of driver
instances.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: serial: stm32: Fix flawed runtime config check

Replace incorrect and limited runtime parity/data bits conditional
configuration with build-time logic using BUILD_ASSERT. This allows for
more flexible UART configurations, while preventing invalid DTS
configurations at build-time.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
- Split into individual commits to assist review

erwango reviewed (changes requested):
Looks ok generally.
The only point is the use of new struct in the device const config struct while in theory runtime updatable parameters should go in data struct.

erwango commented on drivers/serial/uart_stm32.h:
```diff
@@ -13,9 +13,16 @@
 #define ZEPHYR_DRIVERS_SERIAL_UART_STM32_H_
 
 #include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/uart.h>
```
Ok, then maybe do this in a dedicated commit with this explanation

knthm replied:
Sure thing, will do. It's a leftover from the old PR were the reinit function was included in `power.c`.

knthm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1924,12 +1924,24 @@ PINCTRL_DT_INST_DEFINE(index);						\
 static const struct stm32_pclken pclken_##index[] =			\
 					    STM32_DT_INST_CLOCKS(index);\
 									\
+static struct uart_config uart_cfg_##index = {				\
+	.baudrate  = DT_INST_PROP_OR(index, current_speed,		\
+					STM32_UART_DEFAULT_BAUDRATE),	\
+	.parity    = DT_INST_ENUM_IDX_OR(index, parity,			\
+					STM32_UART_DEFAULT_PARITY),	\
+	.stop_bits = DT_INST_ENUM_IDX_OR(index, stop_bits,		\
+					STM32_UART_DEFAULT_STOP_BITS),	\
+	.data_bits = DT_INST_ENUM_IDX_OR(index, data_bits,		\
+					STM32_UART_DEFAULT_DATA_BITS),	\
+	.flow_ctrl = DT_INST_PROP(index, hw_flow_control)		\
+			? UART_CFG_FLOW_CTRL_RTS_CTS			\
+			: UART_CFG_FLOW_CTRL_NONE,			\
+};									\
```
Good point, makes sense now seeing how config is defined const while data is not, I'm still learning the Zephyr conventions bit by bit :wink: 

I like the idea of keeping the `uart_config` as a separate struct with a pointer, so I can shallow copy the contents with a one-liner in `uart_stm32_configure`. Are you opposed to this? I guess having an additional static struct in the driver context could be a considered a bit unclean.

erwango replied:
Keeping uart_config as a separate struct with a pointer looks ok to me if you can store the pointer in data.
Technically storing it in config is correct, but keeping it in data is more consistent with usual practices.

knthm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1939,12 +1951,12 @@ static const struct uart_stm32_config uart_stm32_cfg_##index = {	\
 									\
 static struct uart_stm32_data uart_stm32_data_##index = {		\
-	.baud_rate = DT_INST_PROP(index, current_speed),		\
 	.reset = RESET_DT_SPEC_GET(DT_DRV_INST(index)),			\
```
I could move this over while I'm already making changes to config/data.

erwango replied:
Would be great, txs

knthm replied:
The way the DMA stuff is written, it [modifies its properties](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_stm32.c#L1156-L1209) during runtime. So this would require a significant rework that should probably be a different PR.

knthm committed:
drivers: serial: stm32: Make header file self-contained

Move select include directives from source to header file, as these pull
data types required by the struct definitions contained within it.

This promotes a more modular file structure.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: serial: stm32: Move boot-time config to data

Move reset configuration from uart_stm32_data to
const uart_stm32_config struct, as this is set once at boot and isn't
modified during runtime.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented on drivers/serial/uart_stm32.c:
```diff
@@ -415,9 +415,8 @@ static inline enum uart_config_flow_control uart_stm32_ll2cfg_hwctrl(uint32_t fc
-static int uart_stm32_configure(const struct device *dev,
-				const struct uart_config *cfg)
+static inline void uart_stm32_parameters_set(const struct device *dev,
+						const struct uart_config *cfg)
```
Fixed and removed inline keyword, as function is called more than once.

knthm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1951,6 +1951,109 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+		" has unsupported stop bits configuration");
+/* LPUARTs don't support 1.5 stop bits configurations */
+#elif HAS_LPUART_1
```
Much better, thanks!

knthm commented:
- Add separate commit for include directives move
- Add separate commit for moving reset configuration to config struct
- Move `uart_config` pointer to `uart_stm32_data` struct
- Fix indentation and remove inline keyword from function called more than once
- Simply stop bit `BUILD_ASSERT` check

knthm commented:
- Rebase on main and resolve conflicts

gautierg-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -2014,6 +2014,105 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
+					== UART_CFG_DATA_BITS_5 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_6),	\
```
6-bit payload with a parity bit is possible on STM32 (that's a valid case of LL_USART_DATAWIDTH_7B), so this condition should be added to the check.
Though the function `uart_stm32_cfg2ll_databits` is incomplete because it does not support the case 6-bit data + parity.

knthm replied:
Ahh so parity is included in the `LL_USART_DATAWIDTH` calculation? Looking at the HAL back when I created this PR, this wasn't clear:
```c
#define LL_USART_DATAWIDTH_7B      ...    /*!< 7 bits word length : Start bit, 7 data bits, n stop bits */

```

I'll remove the exclusion of 6-bit data, thanks! Though ideally we should probably check if we have a parity bit defined in the case of 6-bit data.

knthm replied:
Adapted the check to ensure that parity is set if 6-data bits are defined on `LL_USART_DATAWIDTH_7B`.

gautierg-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -2014,6 +2014,105 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
+					== UART_CFG_DATA_BITS_6 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_7),	\
```
Likewise, 7-bit payload + parity is a valid case of LL_USART_DATAWIDTH_8B and should not spring an assert

knthm replied:
Ditto for my comment above.

knthm replied:
Adapted the check to ensure that parity is set if 7-data bits are defined with `LL_USART_DATAWIDTH_8B`.

gautierg-st reviewed (changes requested):
Some comments about the datawidth check, otherwise looks good

erwango commented:
https://github.com/zephyrproject-rtos/zephyr/pull/54354 has been merged. Please rebase

knthm commented:
- Rebase on main and resolve conflicts
